BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsProvider.cs

```diff
@@ -0,0 +1,119 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.SelfDiagnostics
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics.Tracing;
+
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.ConfigString;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Platform;
+
+    using static System.FormattableString;
+
+    /// <summary>
+    /// This class encapsulates parsing and interpreting the self diagnostics configuration string.
+    /// </summary>
+    internal static class SelfDiagnosticsProvider
+    {
+        internal const string SelfDiagnosticsEnvironmentVariable = "APPLICATIONINSIGHTS_SELF_DIAGNOSTICS";
+        internal const string KeyDestination = "Destination";
+        internal const string KeyDirectory = "Directory";
+        internal const string KeyLevel = "Level";
+        internal const string ValueFile = "file";
+        internal const string DefaultDirectory = "%TEMP%";
+        internal const EventLevel DefaultLevel = EventLevel.Verbose;
+
+        /// <summary>
+        /// Parse the environment variable and determine if self diagnostics logging has been enabled. 
+        /// </summary>
+        /// <returns>If the config is valid, will return an instance of the generic class specified. If not, will return null.</returns>
+        /// <exception cref="Exception">
+        /// Throws an exception if the configuration string is invalid. 
+        /// This is expected to crash the application and provide immediate feedback to the end user if the config is invalid.
+        /// </exception>
+        internal static T EvaluateSelfDiagnosticsConfig<T>() where T : ISelfDiagnosticsFileWriter, new()
+        { 
+            try
+            {
+                if (PlatformSingleton.Current.TryGetEnvironmentVariable(SelfDiagnosticsEnvironmentVariable, out string selfDiagnosticsConfigurationString)
+                    && IsFileDiagnosticsEnabled(selfDiagnosticsConfigurationString, out string directory, out string level))
+                {
+                    var instance = new T();
+                    instance.Initialize(level: level, fileDirectory: directory);
+                    return instance;
+                }
+                else
+                {
+                    return default(T);
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new Exception("Failed to parse Self-Diagnostics config string. You must fix or remove this configuration.", ex);
```

Comment:
won't this crash app startup?

Reply:
that was the initial design yes. but I don't want to merge it without having a discussion.

Reply:
The user scenario would be... SDK isn't working as expected. Set EnVar to turn on file logging. If file logging isn't working as expected now a user would have two issues. 

I'm comfortable with this change only because it's meant for troubleshooting scenarios.
We can fail early to let the customer know that the EnVar is set incorrectly. 

If we don't do this, the customer would need something like PerfView to investigate which might not be an option in some Azure environments.

Reply:
While it may be good/bad to prevent AppStartup due to telemetry errors, I'd stick to the existing behaviour for the SDK. We never throw, even when user provides incorrect config. I'll just the same behavior.

Reply:
Probably display a warning message to say the environment variable is ignored?

---
In reply to: [475831878](https://github.com/microsoft/ApplicationInsights-dotnet/pull/1999#discussion_r475831878) [](ancestors = 475831878)

Reply:
@ThomsonTan The question is "where to display"? :D  Server SDKs traditionally have been using EventSource to log that something went bad. But unless user is listening to EventSource using some tools like PerfView/others, there is no way they are going to know something went wrong.

Reply:
I see. If there is no guaranteed way to deliver the warning/error message to users in server SDK, perhaps we could validate the environment variable as early as possible, so the app can only fail during early startup time which the user will be aware of.